components/brave_wallet/renderer/js_ethereum_provider.cc

```diff
@@ -643,4 +653,126 @@ void JSEthereumProvider::MessageEvent(const std::string& subscription_id,
+  auto functionInstance =
+      onProviderRequested->GetFunction(context).ToLocalChecked();
+
+  std::vector<v8::Local<v8::Value>> args;
+  args.push_back(content::V8ValueConverter::Create()->ToV8Value(
+      base::Value("eip6963:requestProvider"), context));
+  args.push_back(std::move(functionInstance));
+
+  CallMethodOfObject(render_frame()->GetWebFrame(), "window",
+                     "addEventListener", std::move(args));
+}
+
+void JSEthereumProvider::AnnounceProvider() {
+  v8::Isolate* isolate = blink::MainThreadIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Context> context =
+      render_frame()->GetWebFrame()->MainWorldScriptContext();
+
+  base::Value::Dict provider_info_value;
+  provider_info_value.Set("rdns", "com.brave.wallet");
+  provider_info_value.Set("uuid", uuid_);
+  provider_info_value.Set(
+      "name", l10n_util::GetStringUTF8(IDS_WALLET_EIP6963_PROVIDER_NAME));
+  provider_info_value.Set("icon", GetBraveWalletImage());
+
+  auto detail = v8::Object::New(isolate);
+  auto info_object = content::V8ValueConverter::Create()->ToV8Value(
+      std::move(provider_info_value), context);
+
+  if (!info_object.As<v8::Object>()
+           ->SetIntegrityLevel(isolate->GetCurrentContext(),
+                               v8::IntegrityLevel::kFrozen)
+           .ToChecked()) {
+    NOTREACHED();
+    return;
+  }
+
+  if (detail
+          ->Set(context,
+                content::V8ValueConverter::Create()->ToV8Value(
+                    base::Value("info"), context),
+                std::move(info_object))
+          .IsNothing()) {
+    NOTREACHED();
+    return;
+  }
+
+  if (detail
+          ->Set(context,
+                content::V8ValueConverter::Create()->ToV8Value(
+                    base::Value("provider"), context),
+                ethereum_provider_proxy_.Get(isolate))
+          .IsNothing()) {
+    NOTREACHED();
+    return;
+  }
+
+  if (detail
+          ->SetIntegrityLevel(isolate->GetCurrentContext(),
+                              v8::IntegrityLevel::kFrozen)
+          .IsNothing()) {
+    NOTREACHED();
+    return;
+  }
+
+  auto event_content = v8::Object::New(isolate);
+  if (event_content
+          ->Set(context,
+                content::V8ValueConverter::Create()->ToV8Value(
+                    base::Value("detail"), context),
+                std::move(detail))
+          .IsNothing()) {
+    NOTREACHED();
+    return;
+  }
+
+  v8::Local<v8::Function> custom_event =
+      GetProperty(context, context->Global(), "CustomEvent")
+          .ToLocalChecked()
+          .As<v8::Function>();
+
+  v8::Local<v8::Value> custom_event_args[] = {
+      content::V8ValueConverter::Create()->ToV8Value(
+          base::Value("eip6963:announceProvider"), context),
+      std::move(event_content)};
+
+  v8::Local<v8::Value> custom_event_value =
+      custom_event->NewInstanceWithSideEffectType(context, 2, custom_event_args)
```

Comment:
why not using `NewInstance` directly? we don't specify type other than `kHasSideEffect`.
And argc should be 1 here, please use `sizeof(custom_event_args)`.